File: python/nothing/battery.py
"""Battery health report for Nothing phones."""

from .device import run
from .models import DeviceInfo

# dumpsys battery → integer health codes
_HEALTH = {
    1: "Unknown",
    2: "Good",
    3: "Overheat",
    4: "Dead",
    5: "Over voltage",
    6: "Unspecified failure",
    7: "Cold",
}

# dumpsys battery → integer status codes
_STATUS = {
    1: "Unknown",
    2: "Charging",
    3: "Discharging",
    4: "Not charging",
    5: "Full",
}

# dumpsys battery → integer plugged codes
_PLUGGED = {
    0: "Not plugged",
    1: "AC",
    2: "USB",
    4: "Wireless",
}


def _parse_dumpsys_battery(output: str) -> dict[str, str]:
    """Return a flat key→value dict from 'dumpsys battery' output."""
    result: dict[str, str] = {}
    for line in output.splitlines():
        if ": " in line:
            key, _, value = line.partition(": ")
            result[key.strip()] = value.strip()
    return result


def _get_cycle_count(serial: str) -> str:
    """Try multiple sources for battery cycle count; return a display string."""
    # 1 — batterystats grep
    r = run(["adb", "-s", serial, "shell",
             "dumpsys batterystats | grep -E 'Charge cycle count'"])
    if r.returncode == 0 and r.stdout.strip():
        for line in r.stdout.strip().splitlines():
            if "Charge cycle count" in line and "=" in line:
                raw = line.split("=")[-1].strip().split()[0]
                try:
                    return str(int(raw))
                except ValueError:
                    pass

    # 2 — sysfs node
    r2 = run(["adb", "-s", serial, "shell",
              "cat /sys/class/power_supply/battery/cycle_count 2>/dev/null"])
    if r2.returncode == 0 and r2.stdout.strip():
        try:
            return str(int(r2.stdout.strip()))
        except ValueError:
            pass

    return "(not available on this kernel)"


def action_battery(device: DeviceInfo) -> None:
    """Display a battery health report for the connected Nothing phone."""
    # ── dumpsys battery ────────────────────────────────────────────────────
    r = run(["adb", "-s", device.serial, "shell", "dumpsys battery"])
    fields = _parse_dumpsys_battery(r.stdout) if r.returncode == 0 else {}

    def _int(key: str) -> int | None:
        try:
            return int(fields[key])
        except (KeyError, ValueError):
            return None

    level_raw   = _int("level")
    status_raw  = _int("status")
    health_raw  = _int("health")
    temp_raw    = _int("temperature")
    voltage_raw = _int("voltage")
    plugged_raw = _int("plugged")

    level_str  = f"{level_raw} %" if level_raw is not None else "not available"
    status_str = _STATUS.get(status_raw, f"unknown ({status_raw})") if status_raw is not None else "not available"
    health_str = _HEALTH.get(health_raw, f"unknown ({health_raw})") if health_raw is not None else "not available"

    # Some devices (e.g. Phone 1) report power source as individual boolean fields
    # instead of a single numeric "plugged" field.
    if plugged_raw is not None:
        plugged_str = _PLUGGED.get(plugged_raw, f"unknown ({plugged_raw})")
    elif fields.get("AC powered") == "true":
        plugged_str = "AC"
    elif fields.get("USB powered") == "true":
        plugged_str = "USB"
    elif fields.get("Wireless powered") == "true":
        plugged_str = "Wireless"
    elif fields.get("Dock powered") == "true":
        plugged_str = "Dock"
    else:
        plugged_str = "Not plugged"

    if temp_raw is not None:
        temp_c = temp_raw / 10
        temp_str = f"{temp_c:.1f} \u00b0C"
    else:
        temp_str = "not available"

    if voltage_raw is not None:
        voltage_v = voltage_raw / 1000
        voltage_str = f"{voltage_v:.2f} V"
    else:
        voltage_str = "not available"

    # ── cycle count ────────────────────────────────────────────────────────
    cycle_str = _get_cycle_count(device.serial)

    # ── battery capacity ───────────────────────────────────────────────────
    # Try charge_full (current max, µAh) then charge_full_design (nominal, µAh).
    # Some MediaTek kernels report charge_full_design in a non-µAh unit — we
    # sanity-check and skip values that would convert to < 500 mAh.
    capacity_str: str | None = None
    for sysfs_node, label in [
        ("charge_full",        "current max"),
        ("charge_full_design", "design"),
    ]:
        r_cap = run(["adb", "-s", device.serial, "shell",
                     f"cat /sys/class/power_supply/battery/{sysfs_node} 2>/dev/null"])
        if r_cap.returncode == 0 and r_cap.stdout.strip():
            try:
                uah = int(r_cap.stdout.strip())
                mah = uah // 1000
                if mah >= 500:          # sanity check: a phone battery is ≥ 500 mAh
                    capacity_str = f"{mah} mAh ({label})"
                    break
            except ValueError:
                pass

    # ── output ────────────────────────────────────────────────────────────
    print(f"\n  Battery Report \u2014 {device.model}\n")
    print(f"  {'Level':<13}: {level_str}")
    print(f"  {'Status':<13}: {status_str}")
    print(f"  {'Health':<13}: {health_str}")
    print(f"  {'Temperature':<13}: {temp_str}")
    print(f"  {'Voltage':<13}: {voltage_str}")
    print(f"  {'Plugged':<13}: {plugged_str}")
    print()
    print(f"  {'Cycle count':<13}: {cycle_str}  (estimated \u2014 varies by kernel)")
    if capacity_str:
        print(f"  {'Capacity':<13}: {capacity_str}")
    print()

File: go/internal/battery/battery.go
// Package battery provides battery health reporting for Nothing phones.
package battery

import (
	"fmt"
	"strings"

	"github.com/Limplom/nothingctl/internal/adb"
)

var healthLabels = map[int]string{
	1: "Unknown",
	2: "Good",
	3: "Overheat",
	4: "Dead",
	5: "Over voltage",
	6: "Unspecified failure",
	7: "Cold",
}

var statusLabels = map[int]string{
	1: "Unknown",
	2: "Charging",
	3: "Discharging",
	4: "Not charging",
	5: "Full",
}

var pluggedLabels = map[int]string{
	0: "Not plugged",
	1: "AC",
	2: "USB",
	4: "Wireless",
}

func parseDumpsysBattery(output string) map[string]string {
	result := make(map[string]string)
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimRight(line, "\r")
		if idx := strings.Index(line, ": "); idx >= 0 {
			key := strings.TrimSpace(line[:idx])
			val := strings.TrimSpace(line[idx+2:])
			result[key] = val
		}
	}
	return result
}

func getCycleCount(serial string) string {
	// Method 1: batterystats
	stdout, _, _ := adb.Run([]string{"adb", "-s", serial, "shell",
		"dumpsys batterystats | grep -E 'Charge cycle count'"})
	for _, line := range strings.Split(stdout, "\n") {
		if strings.Contains(line, "Charge cycle count") && strings.Contains(line, "=") {
			parts := strings.Split(line, "=")
			raw := strings.Fields(parts[len(parts)-1])
			if len(raw) > 0 {
				var n int
				if _, err := fmt.Sscanf(raw[0], "%d", &n); err == nil {
					return fmt.Sprintf("%d", n)
				}
			}
		}
	}
	// Method 2: sysfs
	stdout2, _, _ := adb.Run([]string{"adb", "-s", serial, "shell",
		"cat /sys/class/power_supply/battery/cycle_count 2>/dev/null"})
	raw := strings.TrimSpace(stdout2)
	if raw != "" {
		var n int
		if _, err := fmt.Sscanf(raw, "%d", &n); err == nil {
			return fmt.Sprintf("%d", n)
		}
	}
	return "(not available on this kernel)"
}

func intField(fields map[string]string, key string) (int, bool) {
	v, ok := fields[key]
	if !ok {
		return 0, false
	}
	var n int
	if _, err := fmt.Sscanf(strings.TrimSpace(v), "%d", &n); err != nil {
		return 0, false
	}
	return n, true
}

// ActionBattery displays a battery health report for the connected Nothing phone.
func ActionBattery(serial string) error {
	model := strings.TrimSpace(func() string {
		s, _, _ := adb.Run([]string{"adb", "-s", serial, "shell", "getprop ro.product.model"})
		return s
	}())

	stdout, _, _ := adb.Run([]string{"adb", "-s", serial, "shell", "dumpsys battery"})
	fields := parseDumpsysBattery(stdout)

	levelRaw, hasLevel := intField(fields, "level")
	statusRaw, hasStatus := intField(fields, "status")
	healthRaw, hasHealth := intField(fields, "health")
	tempRaw, hasTemp := intField(fields, "temperature")
	voltageRaw, hasVoltage := intField(fields, "voltage")
	pluggedRaw, hasPlugged := intField(fields, "plugged")

	levelStr := "not available"
	if hasLevel {
		levelStr = fmt.Sprintf("%d %%", levelRaw)
	}
	statusStr := "not available"
	if hasStatus {
		if label, ok := statusLabels[statusRaw]; ok {
			statusStr = label
		} else {
			statusStr = fmt.Sprintf("unknown (%d)", statusRaw)
		}
	}
	healthStr := "not available"
	if hasHealth {
		if label, ok := healthLabels[healthRaw]; ok {
			healthStr = label
		} else {
			healthStr = fmt.Sprintf("unknown (%d)", healthRaw)
		}
	}

	pluggedStr := "Not plugged"
	if hasPlugged {
		if label, ok := pluggedLabels[pluggedRaw]; ok {
			pluggedStr = label
		} else {
			pluggedStr = fmt.Sprintf("unknown (%d)", pluggedRaw)
		}
	} else if fields["AC powered"] == "true" {
		pluggedStr = "AC"
	} else if fields["USB powered"] == "true" {
		pluggedStr = "USB"
	} else if fields["Wireless powered"] == "true" {
		pluggedStr = "Wireless"
	} else if fields["Dock powered"] == "true" {
		pluggedStr = "Dock"
	}

	tempStr := "not available"
	if hasTemp {
		tempStr = fmt.Sprintf("%.1f \u00b0C", float64(tempRaw)/10)
	}
	voltageStr := "not available"
	if hasVoltage {
		voltageStr = fmt.Sprintf("%.2f V", float64(voltageRaw)/1000)
	}

	cycleStr := getCycleCount(serial)

	// Battery capacity
	capacityStr := ""
	for _, node := range []struct{ path, label string }{
		{"/sys/class/power_supply/battery/charge_full", "current max"},
		{"/sys/class/power_supply/battery/charge_full_design", "design"},
	} {
		out, _, _ := adb.Run([]string{"adb", "-s", serial, "shell",
			"cat " + node.path + " 2>/dev/null"})
		raw := strings.TrimSpace(out)
		if raw != "" {
			var uah int64
			if _, err := fmt.Sscanf(raw, "%d", &uah); err == nil {
				mah := uah / 1000
				if mah >= 500 {
					capacityStr = fmt.Sprintf("%d mAh (%s)", mah, node.label)
					break
				}
			}
		}
	}

	fmt.Printf("\n  Battery Report \u2014 %s\n\n", model)
	fmt.Printf("  %-13s: %s\n", "Level", levelStr)
	fmt.Printf("  %-13s: %s\n", "Status", statusStr)
	fmt.Printf("  %-13s: %s\n", "Health", healthStr)
	fmt.Printf("  %-13s: %s\n", "Temperature", tempStr)
	fmt.Printf("  %-13s: %s\n", "Voltage", voltageStr)
	fmt.Printf("  %-13s: %s\n", "Plugged", pluggedStr)
	fmt.Println()
	fmt.Printf("  %-13s: %s  (estimated \u2014 varies by kernel)\n", "Cycle count", cycleStr)
	if capacityStr != "" {
		fmt.Printf("  %-13s: %s\n", "Capacity", capacityStr)
	}
	fmt.Println()
	return nil
}
